ambry-router/src/main/java/com.github.ambry.router/GetOperation.java

```diff
@@ -170,16 +170,22 @@ boolean isOperationComplete() {
 
   /**
    * Set the exception associated with this operation.
-   * A {@link ServerErrorCode#Blob_Deleted} or {@link ServerErrorCode#Blob_Expired} error overrides any other
-   * previously received exception.
+   * First, if current operationException is null, directly set operationException as exception;
+   * Second, if operationException exists, compare ErrorCodes of exception and existing operation Exception depending
+   * on precedence level. An ErrorCode with a smaller precedence level overrides an ErrorCode with a larger precedence
+   * level. Update the operationException if necessary.
    * @param exception the {@link RouterException} to possibly set.
    */
   void setOperationException(Exception exception) {
     if (exception instanceof RouterException) {
       RouterErrorCode routerErrorCode = ((RouterException) exception).getErrorCode();
-      if (operationException.get() == null || routerErrorCode == RouterErrorCode.BlobDeleted
-          || routerErrorCode == RouterErrorCode.BlobExpired) {
+      if (operationException.get() == null) {
         operationException.set(exception);
+      } else {
+        if (getPrecedenceLevel(routerErrorCode) < getPrecedenceLevel(
+            ((RouterException) operationException.get()).getErrorCode())) {
```

Comment:
Isn't this casting and dereferencing unsafe? What if `operationException` is not of type `RouterException`? You would need to have a policy for how precedence/overriding would work if the existing exception is not a `RouterException`. Maybe you leave it alone and only override `RouterException`s?